### PR TITLE
Invoke broker through universal link and check for the new "msauthv3" scheme

### PR DESCRIPTION
**Pull request by oldalton**

### CODE REVIEW

File: IdentityCore/src/util/ios/MSIDAppExtensionUtil.m [hunk 1]
```diff
@@ -68,4 +68,24 @@ + (void)sharedApplicationOpenURL:(NSURL*)url
...
+        return;
+    }
+    
+    dispatch_async( dispatch_get_main_queue(), ^{
+        
+        SEL openURLSelector = @selector(openURL:options:completionHandler:);
+        UIApplication *application = [self sharedApplication];
+        id (*safeOpenURL)(id, SEL, id, id, id) = (void *)[application methodForSelector:openURLSelector];
+        
+        safeOpenURL(application, openURLSelector, url, options, completionHandler);
```
Comment (jasoncoolmax) [inline IdentityCore/src/util/ios/MSIDAppExtensionUtil.m]: could you explain a little bit on why we cannot call openURL directly? :)

File: IdentityCore/src/MSIDConstants.h [hunk 2]
```diff
@@ -81,6 +81,7 @@ extern NSString * _Nonnull const MSID_BROKER_RESUME_DICTIONARY_KEY;
 extern NSString * _Nonnull const MSID_BROKER_SYMMETRIC_KEY_TAG;
 extern NSString * _Nonnull const MSID_BROKER_ADAL_SCHEME;
 extern NSString * _Nonnull const MSID_BROKER_MSAL_SCHEME;
+extern NSString * _Nonnull const MSID_BROKER_NONCE_SCHEME;
```
Comment (antrix1989) [inline IdentityCore/src/MSIDConstants.h]: We might want to rename "nonce".

File: IdentityCore/src/controllers/ios/MSIDBrokerInteractiveController.m [hunk 3]
```diff
@@ -199,17 +207,44 @@ - (void)callBrokerWithRequest:(MSIDBrokerTokenRequest *)brokerRequest
...
 
+- (void)openBrokerWithRequestURL:(NSURL *)requestURL
+{
+    NSDictionary *options = nil;
+    
+    if (self.interactiveParameters.brokerInvocationOptions.isUniversalLink)
+    {
+        // Option for openURL:options:CompletionHandler: only open URL if it is a valid universal link with an application configured to open it
+        // If there is no application configured, or the user disabled using it to open the link, completion handler called with NO
+        if (@available(iOS 10.0, *)) {
```
Comment (antrix1989) [inline IdentityCore/src/controllers/ios/MSIDBrokerInteractiveController.m]: nit: new line

File: IdentityCore/src/parameters/MSIDBrokerInvocationOptions.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDBrokerInvocationOptions : NSObject
+
+@property (nonatomic, readonly) MSIDRequiredBrokerType minRequiredBrokerType;
+@property (nonatomic, readonly) MSIDBrokerProtocolType protocolType;
+@property (nonatomic, readonly) MSIDBrokerAADRequestVersion brokerAADRequestVersion;
+@property (nonatomic, readonly) BOOL isRequiredBrokerPresent;
+@property (nonatomic, readonly) NSString *brokerBaseUrlString;
+@property (nonatomic, readonly) NSString *versionDisplayableName;
```
Comment (antrix1989) [inline IdentityCore/src/parameters/MSIDBrokerInvocationOptions.h]: is this param used only for logging?

File: IdentityCore/src/util/ios/MSIDAppExtensionUtil.m (hunk 1 above)
Comment (oldalton) [reply IdentityCore/src/util/ios/MSIDAppExtensionUtil.m]: Synced offline. Identitycore only allows usage of extension safe APIs. openURL is unsafe API, that's why we're using performSelect in normal openURL. However, performSelector cannot accept 3 parameters.

File: IdentityCore/src/parameters/MSIDBrokerInvocationOptions.h (hunk 4 above)
Comment (oldalton) [reply IdentityCore/src/parameters/MSIDBrokerInvocationOptions.h]: jepp :)

File: IdentityCore/src/controllers/ios/MSIDBrokerInteractiveController.m (hunk 3 above)
Comment (oldalton) [reply IdentityCore/src/controllers/ios/MSIDBrokerInteractiveController.m]: fixed

**Pull request merged**